react/more-react-concepts/refs_and_memoization/refs_and_memoization.md

```diff
@@ -0,0 +1,384 @@
+### Introduction
+
+How can we do DOM manipulations that we usually do in vanilla JavaScript? Is it even possible in React because of its declarative nature? How about memoization? As we all know, when a state has been changed, React will try to re-render the component, which means that it will destroy all local variables not controlled by React and re-execute them. This mental model also works for a regular function that you create, as a component is just a function. Since this is the case, wouldn't it slow down the app if we have a very expensive calculation re-executed with each re-render?
+
+### Lesson Overview
+
+This section contains a general overview of topics you will learn in this lesson.
+
+- Explore `useRef` hook and its use cases
+- Explain memoization and how `useCallback` and `useMemo` can be used
+
+### The useRef hook
+
+The `useRef` hook provides a way to directly access and interact with DOM elements and Child components in React. They are often used when performing imperative actions or accessing specific elements rendered in the DOM. Refs can also *persist* values throughout the component's lifecycle, meaning that the value of the ref will not be destroyed every time a component re-renders. This is very useful when you want to store a value that you want to persist throughout the component's lifecycle without storing it in a state.
```

Comment:
Refs aren't *just* used for storing DOM nodes - in fact, that's one of the lesser used functions that they have. They are used for storing returned IDs from setInterval or setTimeout, date objects, input values that do not need to trigger re-renders, etc. Basically, they act as a catch-all for values that need to be stored but "aren't state".

Reply:
> Refs aren't _just_ used for storing DOM nodes - in fact, that's one of the lesser used functions that they have. They are used for storing returned IDs from setInterval or setTimeout, date objects, input values that do not need to trigger re-renders, etc. Basically, they act as a catch-all for values that need to be stored but "aren't state".

I did write in the end:
> Refs can also *persist* values throughout the component's lifecycle, meaning that the value of the ref will not be destroyed every time a component re-renders. This is very useful when you want to store a value that you want to persist throughout the component's lifecycle without storing it in a state.

But the start could be changed yeah 👍 I've mostly based the premise of this lesson on DOM Manipulation, because of the number of people writing code that they would write in Vanilla JS (in the server). So I focused more on DOM Manipulation. Yeah refs aren't always the best thing to go for and it's more on how they wrote their components, etc.

Do you think I should add another section below the DOM manipulation section? @Lofty-Brambles

Reply:
I'm specifically concerned about the way refs are introduced. We definitely have a section on non-reactive DOM manipulation, so defining refs as an escape hatch [dumbing it down] in this paragraph, would be my way to go.

Reply:
I think it's _slightly_ better to introduce refs & the `useRef` hook in context of DOM manipulation. It provides an easy to understand premise and obviously a great use-case. By the end of the lesson, we can say "oh by the way you can do much more with refs" and give some assignment readings. So I'm content with how the lesson is doing it currently, though I do understand where you're coming from @Lofty-Brambles.

Reply:
@01zulfi o/ ^^

Reply:
Well, it is the *most common* use case which the docs agree to as well. I can agree on changing the first line but it seems like we are re-iterating it in the 2nd paragraph, also I don't think it is a complement, more like an alternative with very strong differences? To which I can refer them to check **only** this table: https://react.dev/learn/referencing-values-with-refs#differences-between-refs-and-state

IMO:

The `useRef` hook offers a valuable capability of managing a value throughout a component's lifecycle that's not needed for rendering. Unlike regular variables that get destroyed on each component re-render, refs allow you to store and "remember" values without relying on state. Unlike state, refs do not trigger re-renders. They're kind of an alternative to `useState` with very important differences. To understand their differences, read **only** the table here [Differences between refs and state](https://react.dev/learn/referencing-values-with-refs#differences-between-refs-and-state). 

They serve multiple purposes such as:
- Storing returned IDs from `setInterval` or `setTimeout`, which you can use *later* for perhaps clearing the timers.
- Storing input values that do not need to trigger re-renders.
- Storing _any_ value without impacting the rendering logic, meaning they aren't necessary for calculating the JSX.
- Providing direct access to DOM elements.
- Facilitating imperative behaviors (e.g. scrolling to an element, focusing on an element, triggering an animation, etc. Basically any kind of DOM manipulation).

By leveraging refs, you can conveniently retain important data throughout the lifespan of a component, ensuring its availability whenever needed. However, treat this hook as an "escape hatch", you won't always need them, use them when you **need** to communicate with external APIs, which is often the case with the browser API.

What do you guys think? @01zulfi @Lofty-Brambles Let me know if you guys still have some suggestions

Reply:
Well, it is the *most common* use case which the docs agree to as well. I can agree on changing the first line but it seems like we are re-iterating the same information in the 2nd paragraph, also I don't think it is a complement, more like an alternative with very strong differences? To which I can refer them to check **only** this table: https://react.dev/learn/referencing-values-with-refs#differences-between-refs-and-state

I can rewrite it to this:

The `useRef` hook offers a valuable capability of managing a value throughout a component's lifecycle that's not needed for rendering. Unlike regular variables that get destroyed on each component re-render, refs allow you to store and "remember" values without relying on state. Unlike state, refs do not trigger re-renders. They're kind of an alternative to `useState` with very important differences. To understand their differences, read **only** the table here [Differences between refs and state](https://react.dev/learn/referencing-values-with-refs#differences-between-refs-and-state). 

They serve multiple purposes such as:
- Storing returned IDs from `setInterval` or `setTimeout`, which you can use *later* for perhaps clearing the timers.
- Storing input values that do not need to trigger re-renders.
- Storing _any_ value without impacting the rendering logic, meaning they aren't necessary for calculating the JSX.
- Providing direct access to DOM elements.
- Facilitating imperative behaviors (e.g. scrolling to an element, focusing on an element, triggering an animation, etc. Basically, DOM manipulation).

However, treat this hook as an "escape hatch", you won't always need them, use them when you **need** to communicate with external APIs, which is often the case with the browser API.

What do you guys think? @01zulfi @Lofty-Brambles Let me know if you guys still have some suggestions

Reply:
Great suggestions @Lofty-Brambles @fcasibu 

I really like the rewritten section, @fcasibu, though I have an objection. Listing out all the possible use cases is going to burden the learner with a lot of information. I don't think we would want that. It's valuable that they discover these use cases on their own during coding & researching. As such, I think the changes @Lofty-Brambles suggested above are sufficient and gets the point across that refs are not only for DOM stuff.